File: livingdoc-fixture-converter/src/main/java/org/livingdoc/fixture/converter/time/AbstractTemporalConverter.java
package org.livingdoc.fixture.converter.time;

import java.lang.reflect.AnnotatedElement;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Optional;

import org.livingdoc.fixture.api.converter.Format;
import org.livingdoc.fixture.api.converter.TypeConverter;
import org.livingdoc.fixture.converter.exceptions.MalformedFormatException;
import org.livingdoc.fixture.converter.exceptions.ValueFormatException;


abstract class AbstractTemporalConverter<T> implements TypeConverter<T> {

    @Override
    public T convert(String value, AnnotatedElement element) throws ValueFormatException, MalformedFormatException {
        DateTimeFormatter formatter = null;
        try {
            formatter = Optional.ofNullable(element)
                .map(e -> e.getAnnotation(Format.class))
                .map(Format::value)
                .map(DateTimeFormatter::ofPattern)
                .orElseGet(this::getDefaultFormatter);
            return parse(value, formatter);
        } catch (DateTimeParseException e) {
            throw new ValueFormatException(value, String.valueOf(formatter), e);
        } catch (IllegalArgumentException e) {
            throw new MalformedFormatException(e);
        }
    }

    abstract DateTimeFormatter getDefaultFormatter();

    abstract T parse(String value, DateTimeFormatter formatter);

}

File: livingdoc-fixture-converter/src/main/java/org/livingdoc/fixture/converter/time/AbstractTemporalConverter.kt
package org.livingdoc.fixture.converter.time

import org.livingdoc.fixture.api.converter.Format
import org.livingdoc.fixture.api.converter.TypeConverter
import org.livingdoc.fixture.converter.exceptions.MalformedFormatException
import org.livingdoc.fixture.converter.exceptions.ValueFormatException
import java.lang.reflect.AnnotatedElement
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.time.temporal.Temporal


abstract class AbstractTemporalConverter<T : Temporal> : TypeConverter<T> {

    @Throws(ValueFormatException::class, MalformedFormatException::class)
    override fun convert(value: String, element: AnnotatedElement?): T {
        val formatter = getDateTimeFormatter(element)
        return parse(value, formatter)
    }

    private fun getDateTimeFormatter(element: AnnotatedElement?): DateTimeFormatter {
        try {
            val customFormatter = element
                    ?.getAnnotation(Format::class.java)
                    ?.value
                    ?.let { DateTimeFormatter.ofPattern(it) }
            return customFormatter ?: defaultFormatter()
        } catch (e: IllegalArgumentException) {
            throw MalformedFormatException(e)
        }
    }

    private fun parse(value: String, formatter: DateTimeFormatter): T {
        try {
            return doParse(value, formatter)
        } catch (e: DateTimeParseException) {
            throw ValueFormatException(value, formatter.toString(), e)
        }
    }

    protected abstract fun defaultFormatter(): DateTimeFormatter

    @Throws(DateTimeParseException::class)
    protected abstract fun doParse(value: String, formatter: DateTimeFormatter): T

}
